### DISCUSSION TITLE
How to use nested map in Move

### DISCUSSION THREAD

**Discussion by 0x-j** [aptos-labs/aptos-developer-discussions#63] [Questions] [answered]
Asked by `skite_in` from discord

How do I create nested maps in Move?

**Comment by 0x-j** [comment]
You can have nested table (which is map in Aptos) in your struct like this
```
    struct Sellers has key, store {
        sellers: Table<address, Table<address, bool>>
    }
```

**Reply by gregnazario** [reply, marked as answer]
There are three types of "Map-like" data structures in Aptos:

`Table` `SmartTable` and `SimpleMap`

There are tradeoffs for each on gas costs, and ease of use.

It depends on your use case.

I did a segment on this on twitter https://x.com/Greg_Nazario/status/1760809623194677725?s=20

But the tl;dr:
1. `SimpleMap` is based in a vector, it doesn't scale > 10k, so if you want a large set you should stay away from it.  Every lookup may iterate through the whole map.
2. `SimpleMap` can't be nested.
3. `Table` is useful, but it's not iterable.  So, if you need to iterate over multiple values, and you don't know what they are ahead of time, I would not use it.
4. `SmartTable` is iterable, and generally more efficient than `Table`

For your use case (I'm assuming some), is that you probably want to be able to iterate over the inner map.

## Nested Tables
I would instead of `Table<address, Table<address, bool>>` probably use:
```move
 SmartTable<address, SmartTable<address, bool>>
```

## Compound Keys
The other way to approach this is to make a compound key and use a single SmartTable:
```move
struct MyKey has key, store {
  outer: address,
  inner: address,
}

SmartTable<MyKey, bool>

```

## SimpleMap in SmartTable
If the inner map is really big, this could be costly from a gas perspective, as it needs to iterate over all values in the map.
```move
 SmartTable<address, SimpleMap<address, bool>>
```

**Comment by StrawberryChocolateFudge** [comment]
I don't want to open a new discussion, so I ask here:
@gregnazario 
I have a struct

```

   struct Hello has store, drop{
    world: SmartTable<address,bool>
   }
   
```
  Then I run `aptos move prove` I get:

```
error: type `smart_table::SmartTable<address, bool>` is missing required ability `drop`

62 │     world: SmartTable<address,bool>
   
   
```
   What to do? How do I add the required ability drop?

If I remove the drop ability  from the struct it does work but then I get errors all over my codebase where the struct is used because it needs to have that ability

**Reply by 0x-j** [reply]
hi move struct requires inner struct to have all abilities from outer struct, since SmartTable is from aptos framework, you cannot change it. I suggest create a 2 structs, one with drop ability and the rest of fields, the other without drop ability and only contains a smart table
